### ISSUE TITLE
Spurious ECONNRESET

### ISSUE THREAD

**Issue by holgerpieta** [containers/podman#10012]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**
/kind bug

**Description**
My Telegram bot in a node-red in a rootless podman container encounters spurious ECONNRESET when trying to send messages.
I cannot 100 % rule out that the problem is on Telegram side, but as the problems began with me switching from Docker to podman and did not resolve itself since then, I think that would be a rather impressive coincidence. So I think something isn't working correctly (or at least not as expected) in podman container networking.

Initially I thought the problem is IPv6, because I noticed that nslookup returned both IPv4 and IPv6 addresses, but when I tried to CURL the api via IPv6, I got no connection. But I could solve that problem by setting --network=slirp4netns:enable_ipv6=true on the pod. So it doesn't seem to be IPv6.

I set up an automatic retry approach that tries to resend the message up to 3 times whenever it fails, which most of the time works, but sometimes it fails 4 times in a row. So, yes, it's spurious but not necessarily rare.

Research in the internet turned up some (old) problems with slirp4netns mishandling INVALID packages, causing connection resets, so I setup iptables to report whenever that happens:
```
Chain INPUT (policy ACCEPT)
target     prot opt source               destination
LOG        all  --  anywhere             anywhere             state INVALID limit: avg 1/sec burst 5 LOG level debug prefix "invalid: "
```
I did see two or three log entries over the course of a couple of weeks but never when the problem happened. And it happens usually multiple times a day, i.e. every couple of messages.

I checked both host and router logs around times when it happend, but none had anything out of the ordinary, e.g. no DHCP renewals or whatnot.

I know that this is not exactly a helpful error description, but it's all I can deliver in the moment. So this bug report is also a cry for help regarding better debug options.
Whatever you think I should try: I will to the best of my abilities.

**Steps to reproduce the issue:**
It's spurious, so whenever I hammer the api by sending a lot of messages, it just refuses to happen.
The only thing I can say from my point: Run a Telegram bot in node-red and have it send a couple of messages per day. Agreeably somewhat unsatifying.

**Describe the results you received:**
Spurious ECONNRESET

**Describe the results you expected:**
Stable connection.

**Additional information you deem important (e.g. issue happens only occasionally):**
None, see above.

**Output of `podman version`:**
```
Version:      3.0.1
API Version:  3.0.0
Go Version:   go1.15.2
Built:        Thu Jan  1 01:00:00 1970
OS/Arch:      linux/arm
```

**Output of `podman info --debug`:**
```
host:
  arch: arm
  buildahVersion: 1.19.4
  cgroupManager: cgroupfs
  cgroupVersion: v1
  conmon:
    package: 'conmon: /usr/libexec/podman/conmon'
    path: /usr/libexec/podman/conmon
    version: 'conmon version 2.0.27, commit: '
  cpus: 4
  distribution:
    distribution: raspbian
    version: "10"
  eventLogger: journald
  hostname: raspberrypi
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100001
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100001
      size: 65536
  kernel: 5.10.17-v8+
  linkmode: dynamic
  memFree: 1026072576
  memTotal: 3979702272
  ociRuntime:
    name: crun
    package: 'crun: /usr/bin/crun'
    path: /usr/bin/crun
    version: |-
      crun version 0.18.1-7931a-dirty
      commit: 7931a1eab0590eff4041c1f74e2844b297c31cea
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +YAJL
  os: linux
  remoteSocket:
    path: /run/user/1000/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    selinuxEnabled: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: 'slirp4netns: /usr/bin/slirp4netns'
    version: |-
      slirp4netns version 1.1.8
      commit: unknown
      libslirp: 4.3.1-git
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.3.3
  swapFree: 103280640
  swapTotal: 104853504
  uptime: 598h 45m 57.24s (Approximately 24.92 days)
registries:
  search:
  - docker.io
  - quay.io
store:
  configFile: /home/pi/.config/containers/storage.conf
  containerStore:
    number: 7
    paused: 0
    running: 7
    stopped: 0
  graphDriverName: overlay
  graphOptions:
    overlay.mount_program:
      Executable: /usr/bin/fuse-overlayfs
      Package: 'fuse-overlayfs: /usr/bin/fuse-overlayfs'
      Version: |-
        fusermount3 version: 3.4.1
        fuse-overlayfs: version 1.4
        FUSE library version 3.4.1
        using FUSE kernel interface version 7.27
  graphRoot: /home/pi/.local/share/containers/storage
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 15
  runRoot: /run/user/1000/containers
  volumePath: /home/pi/.local/share/containers/storage/volumes
version:
  APIVersion: 3.0.0
  Built: 0
  BuiltTime: Thu Jan  1 01:00:00 1970
  GitCommit: ""
  GoVersion: go1.15.2
  OsArch: linux/arm
  Version: 3.0.1
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**
```
podman/unknown 100:3.0.1-2 armhf
```
(Installed via Kubic repos.)

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide?**
Yes, checked and searched the internet.
No, not the newest version: Kubic stopped supporting podman on "old" distros like Debian / Raspbian 10 (will make a bug report about that).

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Raspberry Pi 4 with Raspbian 10 fully updated.

**Comment by holgerpieta** [comment]
Nobody any ideas for further debugging?

I've added nodes to curl the Telegram API on both IPv4 and IPv6 after ECONNRESET. Oddly enough it got 3 ECONNRESET in a row, each time curl worked like a charm. Then the third retry worked and the message got send.
So I'm not sure anymore it's really a podman problem.

**Comment by mheon** [comment]
@Luap99 Aren't you chasing a similar issue in the compose flakes?

**Comment by Luap99** [comment]
@mheon Yes a similar issue but not the same. Here it seems the the port forwarding generally works but sometimes will respond with ECONNRESET. However in the compose CI test the port forwarder setup fails, no retry will make it work.

@holgerpieta you could try creating the container with `--network=slirp4netns:port_handler=slirp4netns`
This will use the a different port forwarder. Also it might be worth to try with a mtu of 1500, `--network=slirp4netns:port_handler=slirp4netns,mtu=1500`. The default port_handler is rootlesskit and the default mtu is 65520.

**Comment by holgerpieta** [comment]
I played around with `ping -4` and `ping -6` on the host and found:
IPv4: Max MTU is 1280
IPv6: Max MTU is 1472
So I not necessarily surprised that it doesn't work nicely from inside a container if the MTU is set to 65520.

However, before I try setting the MTU in the pod, there is one other thing I noticed:
`ping -6` doesn't work from any rootless container in the pod. `ping -4` work and also `curl -6` works, so it's not that there is no IPv6 at all, but just that ICMPv6 doesn't seem to work.
With that, in my understanding, MTU path discovery isn't going to work and things may fail.
So, probably more important question: How do I get ICMPv6 to work in my rootless pod / containers?

I'll try setting the MTU tomorrow and we'll see if that helps.

**Comment by holgerpieta** [comment]
Ah, if you don't test everything, you can go astray quite easily.

Testing against a more trustworthy server from the host (i.e. no podman in the way) I find maximum package sizes with ping and no_fragment flag:
IPv4: 1464, which is 1500 - 20 (IPv4) - 8 (PPPoE)
IPv6: 1444, which is 1500 - 40 (IPv6) - 8 (PPPoE)
Requesting pings without no_fragment works with any package size up to 47352 for IPv4 (where ever that number comes from).
Haven't bothered to find the maximum for IPv6, but it's larger.
So that's totally fine from a network point of view.

Pinging api.telegram.org on IPv6 works totally fine.
But, pinging them on IPv4 only works with maximum package size of 1252. Everything between that and 1464 fails with a timeout.
So the problem is actually that something on the way to api.telegram.org on IPv4 doesn't like packages > 1280 bytes AND filters out ICMP fragment packages.
I guess that means that every IPv4 connection will fail with packages > 1252 bytes. Packages > 1464 bytes will also fail, because they will be fragmented to at least one package > 1252.
Anyone disagrees if I call for a facepalm here?

What I don't quite understand: Why would that cause a ECONNRESET. Shouldn't it cause a time-out?

Anyway, I've set MTU to 1280, so we will see if that works. Though I really do not like fixing other people's network problems by reducing my MTU,

But all that still leaves the question: Why cannot ping6 from inside a container (it just times out)? What do I have to do to be able to?

**Comment by holgerpieta** [comment]
Just setting the MTU didn't work, but I missed setting the porthandler part. It still worked as in reducing the MTU. But it didn't fix the problem itself.
I'll try the porthandler part (plus MTU) over the next day. But I think if that fixes it, the problem actually isn't the MTU.

**Comment by mheon** [comment]
@AkihiroSuda Any ideas here? Or anything that @holgerpieta can provide to help with a diagnosis?

**Comment by AkihiroSuda** [comment]
>  How do I get ICMPv6 to work in my rootless pod / containers?

Unimplemented. (Needs work on https://gitlab.freedesktop.org/slirp/libslirp)

**Comment by holgerpieta** [comment]
Well, I guess it's now time to say how sorry I am for the noise. And how grateful I am for the help, anyway.

I finally did what I should have done much earlier: I ran a tcpdump on the host interface (so before any podman comes into play) and waited for a connection reset. And guess what I found: The TCP connection got reset by the remote host.
The reason for that is as simple as stupid: The Node-Red Telegram bot open a connection to send a message and then, 20 minutes to hours later, when the next message shall be send, it tries to reuse that connection. And the remote host goes like: WTF? Reset!

So, closing this issue now.
